astropy/cosmology/_utils.py

```diff
@@ -140,3 +143,55 @@ def _depr_kws(func: _F, /, kws: tuple[str, ...], since: str) -> _F:
     wrapper = _depr_kws_wrap(func, kws, since)
     functools.update_wrapper(wrapper, func)
     return wrapper
+
+
+class CachedInDictPropertyDescriptor(Generic[R]):
+    """Descriptor for a property that is cached in the instance's dictionary.
+
+    Note that this is a non-data descriptor, not NOT a data descriptor, so
+    after the property is accessed and cached with the same key as the
+    property, the instance's dictionary will have the property value
+    directly, and the descriptor will not be called again, until the
+    property is deleted from the instance's dictionary. See
+    https://docs.python.org/3/howto/descriptor.html#descriptor-protocol.
+    """
+
+    # __slots__ = ("fget", "name")  # TODO: when __doc__ is supported by __slots__
+
+    def __init__(self, fget: Callable[[Cosmology], R]) -> None:
+        self.fget = fget
+        self.__doc__ = fget.__doc__
+
+    def __set_name__(self, cosmo_cls: type[Cosmology], name: str) -> None:
+        self.name: str = name
+
+    @overload
+    def __get__(
+        self, cosmo: None, cosmo_cls: Any
+    ) -> CachedInDictPropertyDescriptor[R]: ...
+
+    @overload
+    def __get__(self, cosmo: Cosmology, cosmo_cls: Any) -> R: ...
+
+    def __get__(
+        self, cosmo: Cosmology | None, cosmo_cls: type[Cosmology] | None
+    ) -> R | CachedInDictPropertyDescriptor[R]:
+        # Accessed from the class, return the descriptor itself
+        if cosmo is None:
+            return self
+
+        # If the property is not in the instance's dictionary, calculate and store it.
+        if self.name not in cosmo.__dict__:
+            cosmo.__dict__[self.name] = self.fget(cosmo)
+
+        # Return the property value from the instance's dictionary
+        # This is only called once, thereafter the property value is accessed directly
+        # from the instance's dictionary.
+        return cosmo.__dict__[self.name]
+
+
+def cached_on_dict_property(
```

Comment:
This looks a lot like a `lazyproperty` except without `__set__` and `__del__` methods. Could we somehow make it work with, say, `lazyproperty(data_descriptor=False)`?

Indeed, can one not just use [`functools.cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property) here?

Reply:
It looks like `cached_property` would work in this case!
In previous refactors (before this PR) it didn't, but in this iteration I think it should.